### ISSUE TITLE
Custom Fold

### ISSUE THREAD

**Issue by Speed7811** [jacobslusser/ScintillaNET#307]
Dear all,

I'm new with Scintilla and wrote my first lexer successfully :) . The lexer can now Syntax Highlight Ini-Files :) .

Now I want to add a custom fold-function by grouping area - like in Visual Studio. Please see attachments.

My question is:

1. How can can I add the fold-symbols to the lines (I dont find any property)
1a) Should I do this in the lexer? Is it recommend?
2. Is it possible to create a hint if an folded-area is collapsed (see attachment 2)

Thank you all!

Kind regards

Rainer
![2](https://cloud.githubusercontent.com/assets/25704388/22853176/78be1914-f04f-11e6-97e6-77d89ce4d846.jpg)
![1](https://cloud.githubusercontent.com/assets/25704388/22853177/78c0d6b8-f04f-11e6-9232-8c77c8678f6a.jpg)

**Comment by jacobslusser** [comment]
Issue #169.

**Comment by Speed7811** [comment]
Hi,

I read it and followed it... Folding is working now but a VLine is not drawn. What did I wrong? 

I entered this code for example:

`
    Scintilla1.SetProperty("fold", "1")

    For intIndex As Int32 = Marker.FolderEnd To Marker.FolderOpen
      Scintilla1.Markers(intIndex).SetForeColor(SystemColors.ControlLightLight)
      Scintilla1.Markers(intIndex).SetBackColor(SystemColors.ControlDark)
    Next


    Scintilla1.Markers(Marker.Folder).Symbol = MarkerSymbol.BoxPlus
    Scintilla1.Markers(Marker.FolderOpen).Symbol = MarkerSymbol.BoxMinus
    Scintilla1.Markers(Marker.FolderEnd).Symbol = MarkerSymbol.BoxPlusConnected
    Scintilla1.Markers(Marker.FolderMidTail).Symbol = MarkerSymbol.TCorner
    Scintilla1.Markers(Marker.FolderOpenMid).Symbol = MarkerSymbol.BoxMinusConnected
    Scintilla1.Markers(Marker.FolderSub).Symbol = MarkerSymbol.VLine
    Scintilla1.Markers(Marker.FolderTail).Symbol = MarkerSymbol.LCorner

    Scintilla1.AutomaticFold = AutomaticFold.Show Or AutomaticFold.Click Or AutomaticFold.Change
    Scintilla1.SetFoldFlags(FoldFlags.LineAfterContracted)


    Scintilla1.Lines(0).FoldLevel = 0
    Scintilla1.Lines(0).FoldLevelFlags = FoldLevelFlags.Header
    Scintilla1.Lines(1).FoldLevel = 1
    Scintilla1.Lines(2).FoldLevel = 1
    Scintilla1.Lines(3).FoldLevel = 1
    Scintilla1.Lines(4).FoldLevel = 0
`

![2](https://cloud.githubusercontent.com/assets/25704388/22861953/f709dc2e-f124-11e6-8819-912d65147456.jpg)
![3](https://cloud.githubusercontent.com/assets/25704388/22861954/f70aaae6-f124-11e6-996f-ab00d47e6510.jpg)

**Comment by Speed7811** [comment]
So... I searched the wiki and got Issue 81 (https://github.com/jacobslusser/ScintillaNET/issues/81). The TO has the same Problem like me. I tried the last 4 hours to get this simple example to work - no success! Folding is working but not VLine is drawn? I don't know why. What did I wrong?

Has anybody a hint for me? Or a code example?

Thanks a lot!

Rainer

**Comment by jacobslusser** [comment]
```cs
// Configure a margin to display folding symbols
scintilla.Margins[2].Type = MarginType.Symbol;
scintilla.Margins[2].Mask = Marker.MaskFolders;
scintilla.Margins[2].Sensitive = true;
scintilla.Margins[2].Width = 20;

// Set colors for all folding markers
for (int i = Marker.FolderEnd; i <= Marker.FolderOpen; i++)
{
    scintilla.Markers[i].SetForeColor(SystemColors.Control);
    scintilla.Markers[i].SetBackColor(SystemColors.ControlDark);
}

// Configure folding markers with respective symbols
scintilla.Markers[Marker.FolderEnd].Symbol = MarkerSymbol.BoxPlusConnected;
scintilla.Markers[Marker.FolderOpenMid].Symbol = MarkerSymbol.BoxMinusConnected;
scintilla.Markers[Marker.FolderMidTail].Symbol = MarkerSymbol.TCorner;
scintilla.Markers[Marker.FolderTail].Symbol = MarkerSymbol.LCorner;
scintilla.Markers[Marker.FolderSub].Symbol = MarkerSymbol.VLine;
scintilla.Markers[Marker.Folder].Symbol = MarkerSymbol.BoxPlus;
scintilla.Markers[Marker.FolderOpen].Symbol = MarkerSymbol.BoxMinus;

// Insert some text and fold data
scintilla.Text = "function foldingRules() {\r\n    console.log('Folding Rules!');\r\n}\r\n";
scintilla.Lines[0].FoldLevelFlags = FoldLevelFlags.Header;
var foldLevel = scintilla.Lines[0].FoldLevel;
scintilla.Lines[1].FoldLevel = ++foldLevel;
scintilla.Lines[2].FoldLevel = foldLevel;
scintilla.Lines[3].FoldLevel = --foldLevel;

scintilla.MarginClick += (s, mcea) =>
{
    // Toggle the fold when clicking
    var line = scintilla.LineFromPosition(mcea.Position);
    scintilla.Lines[line].ToggleFold();
};

```

Be aware that fold levels do not start at `0`. They start at `1024`.

**Comment by jacobslusser** [comment]
If there are any additional questions we can reopen this ticket.

**Comment by cole-rob** [comment]
Your example works well - for the most part.
How do I keep the folding markers to one margin?  I am seeing the marker in two margins (one should be for breakpoints and the other for the folding).
![image](https://user-images.githubusercontent.com/43245000/93219028-e88ceb00-f738-11ea-957f-630c42624f8f.png)
The breakpoints are in only one margin.

I am using C#
I set up the margin as:
// Configure a margin to display folding symbols
            scintilla1.Margins[2].Type = MarginType.Symbol;
            scintilla1.Margins[2].Mask = Marker.MaskFolders;
            scintilla1.Margins[2].Sensitive = true;
            scintilla1.Margins[2].Width = 20;

My MarginClick function has:
if (e.Margin == 2)
            {
                // Toggle the fold when clicking
                var line = scintilla1.LineFromPosition(e.Position);
                scintilla1.Lines[line].ToggleFold();
            }

**Comment by cole-rob** [comment]
I will answer my own question for the sake of those who are looking at this post.  I revisited this issue after a couple of days.

The problem is not in the definition of the folding margin - the problem is in the definition of my breakpoint margin.
I followed an example for setting up breakpoint markers that used:
    scintilla1.Margins[1].Mask = Marker.MaskAll;
This made ALL markers show up in the margin - including the folding symbols.

I found a [good article on StackOverflow that explains setting up masks in Scintilla](https://stackoverflow.com/questions/27060255/how-to-set-margin-masks-in-scintilla).

I set up my breakpoint marker in marker index 3, so my masking line now looks like:
    scintilla1.Margins[1].Mask = (1 << 3);    // This is simplified with constants removed for clarity.
